swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -447,81 +368,106 @@ extension Adapter: CallbackHandlerDelegate {
     }
   }
 
   public func onUpdateResources(resourceList: String) {
     workQueue.async { [weak self] in
       guard let self = self else { return }
 
-      self.logger.log("Adapter.onUpdateResources")
-      let jsonString = resourceList
-      guard let jsonData = jsonString.data(using: .utf8) else {
-        return
-      }
-      guard let networkResources = try? JSONDecoder().decode([NetworkResource].self, from: jsonData)
-      else {
-        return
-      }
+      logger.log("Adapter.onUpdateResources")
 
-      // Note down the resources
-      self.displayableResources.update(resources: networkResources.map { $0.displayableResource })
+      let networkResources = try! JSONDecoder().decode(
+        [NetworkResource].self, from: resourceList.data(using: .utf8)!)
 
-      // Update DNS in case resource domains is changing
-      guard let networkSettings = self.networkSettings else {
-        self.logger.error("Adapter.onUpdateResources: No network settings")
-        return
-      }
-      let updatedResourceDomains = networkResources.compactMap { $0.resourceLocation.domain }
-      networkSettings.setResourceDomains(updatedResourceDomains)
-      if case .tunnelReady = self.state {
-        networkSettings.apply(on: packetTunnelProvider, logger: self.logger, completionHandler: nil)
-      }
+      // Update resource list
+      displayableResources.update(resources: networkResources.map { $0.displayableResource })
     }
   }
 
   public func onDisconnect(error: String) {
     workQueue.async { [weak self] in
       guard let self = self else { return }
 
-      self.logger.error(
-        "Connlib disconnected with unrecoverable error: \(error)")
-      switch self.state {
-      case .stoppingTunnel(session: _, let onStopped):
-        onStopped?()
-        self.state = .stoppedTunnel
-      case .stoppingTunnelTemporarily(session: _, let onStopped):
-        onStopped?()
-        self.state = .stoppedTunnel
-      case .stoppedTunnel:
-        // This should not happen
-        break
-      case .stoppedTunnelTemporarily:
-        self.state = .stoppedTunnel
-      default:
-        packetTunnelProvider?.handleTunnelShutdown(
-          dueTo: .connlibDisconnected,
-          errorMessage: error)
-        self.packetTunnelProvider?.cancelTunnelWithError(
-          AdapterError.connlibFatalError(error))
-        self.state = .stoppedTunnel
+      logger.log("\(#function)")
+
+      // Unexpected disconnect initiated by connlib. Typically for 401s.
+      logger.error("Connlib disconnected with unrecoverable error: \(error)")
+      stop()
+      stopProvider(error: error)
+    }
+  }
+
+  private func stopProvider(error: String) {
+    Task {
+      var reason: NEProviderStopReason = .connectionFailed
+
+      // HACK: Define more connlib error types across the FFI so we can switch on them
+      // directly and not just sign the user out every time this is called.
+      if error.contains("401 Unauthorized") {
```

Comment:
FYI I now explicitly check for a 401 Unauthorized before clearing the token. Otherwise it just signs the user out quietly.

Reply:
I guess we can refactor this to a typed API once we move to UniFFI!

Reply:
does this matter for other Clients? I don't think I check for 401 anywhere.